### PR TITLE
[TFL] Fix BatchMatMul constant RHS.

### LINKED ISSUE #46724: `tf.matmul` and `tf.tensordot` behave different in converted concrete function in TensorFlowLite
### 1. System information

- OS Platform and Distribution: MacOS 10.15 and Ubuntu 18.04 LTS on Colab machine
- TensorFlow installation (pip package or built from source): pip
- TensorFlow library (version, if pip package or github SHA, if built from source): `tensorflow                    2.4.0`

### 2. Code

This notebook demonstrates the bug with the simplest example I came up with.

https://colab.research.google.com/gist/ebraraktas/ab87170deb38eae979b37795015e44bc

### 3. Failure after conversion

I implemented RFFT for TFLite using `tf.matmul` and saved the module concrete function. But invoking saved tflite model repeatedly returns different results. However, replacing `tf.matmul` with `tf.tensordot` fixes the strange behavior. Therefore, I have prepared the notebook above to demonstrate the bug. I have realized interesting cases which change the behavior:

- If negative sign is removed from output returned from `DummyMatmul` or `DummyTensordot` (`result` variable), outputs are same
- If we use `tf.Module` directly, outputs are same (colab demo shows it)
- Somehow, size of the right hand side matrix matters (colab demo shows it)
- Difference occurs after first iteration (colab demo shows it), and for some inputs it gets larger with every iteration

### PR DESCRIPTION
**Pull request by WindQAQ**
Fixes #46724. For constant RHS and !adj_y, transpose of RHS only occurs once, so the allocation should be persistent across runs. Old commit does this, but the allocation is accidentally overridden to kTfLiteArenaRw for constant RHS.

https://github.com/tensorflow/tensorflow/blob/7c0a3f12dd00b0aeb9d2a91d157b0e9d1a4c1250/tensorflow/lite/kernels/batch_matmul.cc#L193-L199

/cc @abattery for visibility.

### CODE REVIEW

**Comment by thaink** [comment]
Verify that this fixes the issue.

File: tensorflow/lite/kernels/batch_matmul_test.cc [hunk 1]
```diff
@@ -279,6 +279,67 @@ INSTANTIATE_TEST_SUITE_P(
...
+    rhs_id_ = AddConstInput<float>(TensorType_FLOAT32, rhs_data, rhs_shape);
+    matmul_output_id_ = AddOutput(lhs.type);
+    std::vector<int> matmul_inputs{lhs_id_, rhs_id_};
+    std::vector<int> matmul_outputs{matmul_output_id_};
+    AddBuiltinOp(BuiltinOperator_BATCH_MATMUL,
+                 BuiltinOptions_BatchMatMulOptions,
+                 CreateBatchMatMulOptions(builder_, adj_x, adj_y).Union(),
+                 matmul_inputs, matmul_outputs);
+
+    // This is just a dummy op to check if the values of transposed RHS are
```
Comment (thaink) [inline tensorflow/lite/kernels/batch_matmul_test.cc:298]: Can you explain why the neg op help here?

Comment (WindQAQ) [reply tensorflow/lite/kernels/batch_matmul_test.cc:298]: It's somehow tricky. When there is no following ops (not limited to neg), temporary allocation with `kTfLiteArenaRw` type tends to get the same piece of memory across each evaluation and no other ops will modify values at that memory address (cause no other memory allocation). Therefore, the values of rhs are not polluted without neg op, and the next evaluation can also get the same piece of memory, so the results are correct.

https://colab.research.google.com/drive/1eFVszZZN-NcR64X_VHns0w-l3Vwci0CY?usp=sharing

Here is the colab link from original post. The values the last tensor (temp tensor, transposed rhs) become

```
[[-51. -84.]
 [  3.   6.]
 [  7.   9.]]
```

where -51, -84 are the last two elements of `Identity`. While without neg op, everything works smoothly.

Comment (thaink) [reply tensorflow/lite/kernels/batch_matmul_test.cc:298]: Acked. Thanks for explanation.  May be it worths a comment then.

**Review by thaink — APPROVED** [review]

File: tensorflow/lite/kernels/batch_matmul_test.cc (hunk 1 above)
Comment (WindQAQ) [reply tensorflow/lite/kernels/batch_matmul_test.cc]: Updated comments. See if it's better! Thanks!

**Review by thaink — APPROVED** [review]

**Pull request merged**